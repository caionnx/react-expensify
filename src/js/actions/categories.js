import database from '../firebase/firebase'

// REMOVE_CATEGORY
export const removeCategory = (id) => ({
  type: 'REMOVE_CATEGORY',
  id
})

export const startRemoveCategory = ({ id } = {}) => (dispatch, getState) => {
  const uid = getState().auth.uid

  return database.ref(`users/${uid}/categories/${id}`).remove().then(() => {
    dispatch(removeCategory(id))
  })
}

// ADD_CATEGORY
export const addCategory = (category) => ({
  type: 'ADD_CATEGORY',
  category
})

export const startAddCategory = (category = {}) => (dispatch, getState) => {
  const uid = getState().auth.uid
  const { id, value } = category

  return database.ref(`users/${uid}/categories/${id}`).set(value).then(() => {
    dispatch(addCategory(category))
  })
}

export const startRegisterCategories = (categories, uid) => {
  return database.ref(`users/${uid}/categories`).set({...categories})
}

// SET_CATEGORIES
export const setCategories = (categories) => ({
  type: 'SET_CATEGORIES',
  categories
})

export const startSetCategories = () => (dispatch, getState) => {
  const uid = getState().auth.uid
  const transformToArray = (object) => {
    const arr = []
    object.forEach(el => {
      arr.push({
        id: el.key,
        value: el.val()
      })
    })

    return arr
  }

  return database.ref(`users/${uid}/categories`).once('value').then((snapshot) => {
    let categories
    if (snapshot.hasChildren()) {
      categories = transformToArray(snapshot)
      dispatch(setCategories(categories))
    } else {
      database.ref(`categories`).once('value').then((defaultSnapshot) => {
        categories = transformToArray(defaultSnapshot)
        dispatch(setCategories(categories))

        return defaultSnapshot.hasChildren() &&
          startRegisterCategories(defaultSnapshot.val(), uid)
      })
    }
  })
}
