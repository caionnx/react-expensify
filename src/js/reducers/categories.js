export default (state = [], action) => {
  switch (action.type) {
    case 'SET_CATEGORIES':
      return action.categories
    case 'REMOVE_CATEGORY':
      return state.filter(({ id }) => id !== action.id)
    default:
      return state
  }
}
