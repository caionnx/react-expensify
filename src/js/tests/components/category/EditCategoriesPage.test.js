import React from 'react'
import { shallow } from 'enzyme'
import {
  EditCategoriesPage,
  default as ConnectedEditCategoriesPage
} from '../../../components/category/EditCategoriesPage'
import categories from '../../fixtures/categories'
import configureMockStore from 'redux-mock-store'
const mockStore = configureMockStore()

let wrapper, fillCategories, startAddDefaultCategories, startRemoveCategory

beforeEach(() => {
  fillCategories = jest.fn()
  startRemoveCategory = jest.fn(() => Promise.resolve(true))
  startAddDefaultCategories = jest.fn()
  wrapper = shallow(
    <EditCategoriesPage
      fillCategories={fillCategories}
      startRemoveCategory={startRemoveCategory}
      startAddDefaultCategories={startAddDefaultCategories}
      categories={categories} />
  )
})

test('should render EditCategoriesPage correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should open and close modal of remove category', () => {
  const category = wrapper.find('EditCategoryItem').at(0)
  category.props().buttonAction(categories[0].id)
  wrapper.find('Modal').props().onClose()

  expect(wrapper.state('removeCategoryModal')).toBe(false)
})

test('should open modal of remove category and then start removing it', (done) => {
  const category = wrapper.find('EditCategoryItem').at(0)
  category.props().buttonAction(categories[0].id)

  expect(wrapper.state('removeCategoryModal')).toBe(true)

  wrapper.find('Modal').props().onConfirm()

  setTimeout(() => {
    expect(startRemoveCategory).toHaveBeenCalledWith({ id: categories[0].id })
    expect(wrapper.state('removeCategoryModal')).toBe(false)
    done()
  }, 500)
})

test('should render AddStandardCategories if has not categories', () => {
  wrapper = shallow(
    <EditCategoriesPage
      fillCategories={fillCategories}
      startRemoveCategory={startRemoveCategory}
      startAddDefaultCategories={startAddDefaultCategories}
      categories={[]} />
  )
  const AddStandardCategoriesComponent = wrapper.find('AddStandardCategories')
  AddStandardCategoriesComponent.props().buttonAction() // Trigger component button action

  expect(AddStandardCategoriesComponent).toHaveLength(1)
  expect(startAddDefaultCategories).toHaveBeenCalled()
})

test('should have mapStateToProps and mapDispatchToProps', () => {
  const store = mockStore({ categories })
  wrapper = shallow(<ConnectedEditCategoriesPage store={store} />)

  expect(wrapper.prop('categories')).toEqual(categories)
  expect(wrapper.prop('fillCategories')).toBeInstanceOf(Function)
  expect(wrapper.prop('startRemoveCategory')).toBeInstanceOf(Function)
  expect(wrapper.prop('startAddDefaultCategories')).toBeInstanceOf(Function)
})
