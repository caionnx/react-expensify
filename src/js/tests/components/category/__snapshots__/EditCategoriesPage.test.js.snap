// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditCategoriesPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit your categories
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <div
      className="list-header"
    >
      Categories
    </div>
    <div
      className="list-body"
    >
      <EditCategoryItem
        buttonAction={[Function]}
        category={
          Object {
            "id": "shopping",
            "value": "Shopping",
          }
        }
        key="shopping"
      />
      <EditCategoryItem
        buttonAction={[Function]}
        category={
          Object {
            "id": "sports",
            "value": "Sports",
          }
        }
        key="sports"
      />
    </div>
    <Modal
      body="Are you sure you want to remove this category?"
      cancelText="Cancel"
      className="modal"
      confirmText="Remove"
      isOpen={false}
      label="Remove category"
      onClose={[Function]}
      onConfirm={[Function]}
      thunk={true}
    />
    <Connect(AddCategoryForm) />
  </div>
</div>
`;
